src/redisearch_rs/rlookup/src/load_document/mod.rs

```diff
@@ -0,0 +1,545 @@
+/// depending on the document type and the loading mode:
+///
+/// 1. Hash document, load all fields, Scan-Cursor-API: This code path is fully implemented in Rust, see `[hash::get_all_scan`]
+/// 2. Hash document, load all fields, fallback Call-API: This code path is fully implemented in Rust, see `[hash::get_all_fallback`].
+/// 3. JSON document, load all fields: This code path is not yet ported to Rust and still uses the C implementation via FFI, see MOD-11050.
+/// 4. Individual keys (either hash or JSON document): This code path is not yet ported to Rust and still uses the C implementation via FFI, see MOD-11051.
+///
+/// ### Hash Document, Load All Fields, Scan-Cursor-API
+///
+/// Populate the provided `dst_row` by loading all field-value pairs from the [Redis hash] with the key given in `options.dmd.keyPtr`.
+///
+/// Internally either a scan cursor API or a fallback Call API is used, depending on the Redis version and deployment type. In the CRDT
+/// deployment the scan cursor API is not available, so we always use the fallback Call API. The scan cursor API is available
+/// from Redis v6.0.6 and above.
+///
+/// [Redis hash]: https://redis.io/docs/latest/develop/data-types/hashes/
+///
+/// ### Hash Document, Load All Fields, Fallback Call-API
+///
+/// Populate the provided `dst_row` by loading all field-value pairs from the [Redis hash] with the key given in `key_str`
+///
+/// HGETALL implementation using the scan cursor API that is available
+/// from Redis v6.0.6 and above.
+///
+/// [Redis hash]: https://redis.io/docs/latest/develop/data-types/hashes/
+///
+/// ### JSON Document, Load All Fields
+///
+/// Calls in the C code to load all fields from a JSON document.
+///
+/// ### Individual Keys (Either Hash or JSON Document)
+///
+/// Calls in the C code to load individual keys from either a hash or JSON document.
+///
+/// ## References
+///
+/// [Redis hash]: https://redis.io/docs/latest/develop/data-types/hashes/
+pub fn load_document<'a>(
+    lookup: &mut RLookup<'a>,
+    dst_row: &mut RLookupRow<'a, RSValueFFI>,
+    options: &LoadDocumentOptions<'a>,
+) -> Result<(), LoadDocumentError> {
+    todo!("in follow-up PRs:");
+    /*
+    let context = hash::LoadDocumentImpl {
+        ctx: redis_module::Context::dummy(),
+    };
+    load_document_int(lookup, dst_row, options, &context)
+    */
+}
+
+// the following function is only used internally to allow for mocking out the context in tests
+#[inline(always)]
+fn load_document_int<'a, C>(
+    lookup: &mut RLookup<'a>,
+    dst_row: &mut RLookupRow<'a, C::V>,
+    options: &LoadDocumentOptions<'a, C::V>,
+    context: &C,
+) -> Result<(), LoadDocumentError>
+where
+    C: LoadDocumentContext,
+{
+    dst_row.set_sorting_vector(options.sorting_vector);
+
+    match options.mode {
+        RLookupLoadMode::AllKeys => {
+            if options.document_type == DocumentType::Hash {
```

Comment:
What about `Unsupported`? Would get to the `else` block as well.
What about a match expression? (which is exhaustive)